README
NAME
    Catalyst::Plugin::Session::Store::MongoDB - MongoDB session store for
    Catalyst

SYNOPSIS
    In your MyApp.pm:

      use Catalyst qw/
        Session
        Session::Store::MongoDB
        Session::State::Cookie # or similar
      /;

    and in your MyApp.conf

      <Plugin::Session>
        <client_options>             # defaults to letting MongoDB::MongoClient
          host mongodb://foo:27017   # use its own defaults
          timeout 10000
          ...
        </client_options>
        dbname test                  # defaults to catalyst
        collectionname s2            # defaults to session
      </Plugin::Session>

    Then you can use it as usual:

      $c->session->{foo} = 'bar'; # will be saved

DESCRIPTION
    "Catalyst::Plugin::Session::Store::MongoDB" is a session storage plugin
    using MongoDB (<http://www.mongodb.org>) as its backend.

USAGE
    Expired Sessions
        This store automatically deletes sessions when they expire.
        Additionally it implements the optional delete_expired_sessions()
        method.

AUTHOR
      Ronald J Kimball, <rjk@tamias.net>

    Previous Authors

      Stefan Völkel, <bd@bc-bd.org> <http://bc-bd.org>
      Cory G Watson, <gphat at cpan.org>

LICENSE
    This program is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License v2 as published by the
    Free Software Foundation; or the Artistic License.



Changes
Revision history for Perl module Catalyst::Plugin::Session::Store::MongoDB

0.02 Wed Jun  2 16:25:29 CEST 2010
    - complete rewrite by "Stefan Völkel <bd@bc-bd.org>"

0.01 Thu May 27 12:49:18 2010
    - first version by "Cory G Watson <gphat at cpan.org>"


t/lib/Fixture.pm
package Fixture;
use Moose;
use namespace::autoclean;

use String::Random;

has 'collection' => (
  isa => 'MongoDB::Collection',
  is => 'ro',
  lazy_build => 1,
);

has 'store' => (
  isa => 'Catalyst::Plugin::Session::Store::MongoDB',
  is => 'ro',
  lazy_build => 1,
);

has '_cleanup' => (
  isa => 'ArrayRef',
  is => 'rw',
);

has '_random' => (
  isa => 'String::Random',
  is => 'ro',
  lazy_build => 1,
);

sub _build_collection {
  my ($self) = @_;

  return $self->store->_collection;
}

sub _build_store {
  return Catalyst::Plugin::Session::Store::MongoDB->new(
    connection => {
      host => $ENV{MONGODB_HOST},
      port => $ENV{MONGODB_PORT},
    },
    dbname => $ENV{TEST_DB},
    collectionname => $ENV{TEST_COLLECTION},
  );
}

sub _build__random {
  return String::Random->new();
}

sub setup {
  my ($self) = @_;

  my @needs = qw/MONGODB_HOST MONGODB_PORT TEST_DB TEST_COLLECTION/;
  map {
    return 'Must set '.join(",", @needs).' environment variables'
      unless(defined($ENV{$_}));
  } @needs;

  return undef;
}

sub teardown {
  my ($self) = @_;

  $self->store->delete_session_data($_) foreach (@{$self->{_cleanup}});
}

sub new_id {
  my ($self) = @_;

  my ($id, $found);
  do {
    $id = sprintf("%i", int(rand(1000000)));
    $found = $self->collection->find_one({ _id => $id });
  } while ($found);

  push(@{$self->{_cleanup}}, $id);
  
  return $id;
}

sub new_data {
  my ($self) = @_;

  return $self->_random->randpattern("..............................");
}

1;


lib/Catalyst/Plugin/Session/Store/MongoDB.pm
package Catalyst::Plugin::Session::Store::MongoDB;
use strict;
use warnings;

our $VERSION = '0.02';

use Moose;
use namespace::autoclean;

use MongoDB::MongoClient;
use Data::Dumper;

BEGIN { extends 'Catalyst::Plugin::Session::Store' }

has client_options => (
  isa => 'HashRef',
  is => 'ro',
  lazy_build => 1,
);

# deprecated
has hostname => (
  isa => 'Str',
  is => 'ro',
  lazy_build => 1,
);

# deprecated
has port => (
  isa => 'Int',
  is => 'ro',
  lazy_build => 1,
);

has dbname => (
  isa => 'Str',
  is => 'ro',
  lazy_build => 1,
);

has collectionname => (
  isa => 'Str',
  is => 'ro',
  lazy_build => 1,
);

has '_collection' => (
  isa => 'MongoDB::Collection',
  is => 'ro',
  lazy_build => 1,
);

has '_connection' => (
  isa => 'MongoDB::MongoClient',
  is => 'ro',
  lazy_build => 1,
);

has '_db' => (
  isa => 'MongoDB::Database',
  is => 'ro',
  lazy_build => 1,
);

sub _cfg_or_default {
  my ($self, $name, $default) = @_;

  # _session_plugin_config is not present for tests
  if ($self->can('_session_plugin_config')) {
    my $cfg = $self->_session_plugin_config;
    return $cfg->{$name} || $default;
  } else {
    return $default;
  }
}

sub _build_client_options {
  my ($self) = @_;
  return $self->_cfg_or_default('client_options', {
    host => $self->hostname,
    port => $self->port,
  });
}

sub _build_hostname {
  my ($self) = @_;
  return $self->_cfg_or_default('hostname', 'localhost');
}

sub _build_port {
  my ($self) = @_;
  return $self->_cfg_or_default('port', 27017);
}

sub _build_dbname {
  my ($self) = @_;
  return $self->_cfg_or_default('dbname', 'catalyst');
}

sub _build_collectionname {
  my ($self) = @_;
  return $self->_cfg_or_default('collectionname', 'session');
}

sub _build__collection {
  my ($self) = @_;

  return $self->_db->get_collection($self->collectionname);
}

sub _build__connection {
  my ($self) = @_;

  return MongoDB::MongoClient->new($self->client_options);
}

sub _build__db {
  my ($self) = @_;

  return $self->_connection->get_database($self->dbname);
}

sub _serialize {
  my ($self, $data) = @_;

  my $d = Data::Dumper->new([ $data ]);

  return $d->Indent(0)->Purity(1)->Terse(1)->Quotekeys(0)->Dump;
}

sub get_session_data {
  my ($self, $key) = @_;

  my ($prefix, $id) = split(/:/, $key);

  my $found = $self->_collection->find_one({ _id => $id },
    { $prefix => 1, 'expires' => 1 });

  return undef unless $found;

  if ($found->{expires} && time() > $found->{expires}) {
    $self->delete_session_data($id);
    return undef;
  }

  return eval($found->{$prefix});
}

sub store_session_data {
  my ($self, $key, $data) = @_;

  my ($prefix, $id) = split(/:/, $key);

  # we need to not serialize the expires date, since it comes in as an
  # integer and we need to preserve that in order to be able to use
  # mongodb's '$lt' function in delete_expired_sessions()
  my $serialized;
  if ($prefix =~ /^expires$/) {
    $serialized = $data;
  } else {
    $serialized = $self->_serialize($data);
  }

  $self->_collection->update({ _id => $id },
    { '$set' => { $prefix => $serialized } }, { upsert => 1 });
}

sub delete_session_data {
  my ($self, $key) = @_;

  my ($prefix, $id) = split(/:/, $key);

  my $found = $self->_collection->find_one({ _id => $id });
  return unless $found;

  if (exists($found->{$prefix})) {
    if ((scalar(keys(%$found))) > 2) {
      $self->_collection->update({ _id => $id },
        { '$unset' => { $prefix => 1 }} );
      return;
    } else {
      $self->_collection->remove({ _id => $id });
    }
  }
}

sub delete_expired_sessions {
  my ($self) = @_;

  $self->_collection->remove({ 'expires' => { '$lt' => time() } });
}

__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

Catalyst::Plugin::Session::Store::MongoDB - MongoDB session store for Catalyst

=head1 SYNOPSIS

In your MyApp.pm:

  use Catalyst qw/
    Session
    Session::Store::MongoDB
    Session::State::Cookie # or similar
  /;

and in your MyApp.conf

  <Plugin::Session>
    <client_options>             # defaults to letting MongoDB::MongoClient
      host mongodb://foo:27017   # use its own defaults
      timeout 10000
      ...
    </client_options>
    dbname test                  # defaults to catalyst
    collectionname s2            # defaults to session
  </Plugin::Session>

Then you can use it as usual:

  $c->session->{foo} = 'bar'; # will be saved

=head1 DESCRIPTION

C<Catalyst::Plugin::Session::Store::MongoDB> is a session storage plugin using
MongoDB (L<http://www.mongodb.org>) as its backend.

=head1 USAGE

=over 4

=item B<Expired Sessions>

This store automatically deletes sessions when they expire. Additionally it
implements the optional delete_expired_sessions() method.

=back

=head1 AUTHOR

  Ronald J Kimball, <rjk@tamias.net>

Previous Authors

  Stefan Völkel, <bd@bc-bd.org> <http://bc-bd.org>
  Cory G Watson, <gphat at cpan.org>

=head1 LICENSE

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License v2 as published
by the Free Software Foundation; or the Artistic License.

=cut
